src/vmspawn/vmspawn.c

```diff
@@ -585,15 +599,86 @@ static int find_initrd(char **ret_initrd) {
         return -ESRCH;
 }
 
+static int start_virtiofsd(sd_bus *bus, const char *scope, const char *directory, char **ret_state_tempdir, char **ret_sock_name) {
+        _cleanup_(rm_rf_physical_and_freep) char *state = NULL;
+        _cleanup_strv_free_ char **cmdline = NULL, **cleanup = NULL;
+        _cleanup_free_ char *virtiofsd = NULL, *sock_path = NULL, *sock_name = NULL, *rm_path = NULL, *unit_name = NULL;
+        int r;
+
+        assert(bus);
+        assert(scope);
+        assert(directory);
+        assert(ret_state_tempdir);
+        assert(ret_sock_name);
+
+        r = find_executable("virtiofsd", &virtiofsd);
+        if (r < 0 && r != -ENOENT)
+                return log_error_errno(r, "Error while searching for virtiofsd: %m");
+
+        if (!virtiofsd) {
+                FOREACH_STRING(dir, "/usr/libexec/virtiofsd", "/usr/libexec/virtiofsd") {
+                        if (access(dir, X_OK) >= 0) {
+                                virtiofsd = strdup(dir);
+                                if (!virtiofsd)
+                                        return log_oom();
+                                break;
+                        }
+
+                        if (!IN_SET(errno, ENOENT, EACCES))
+                                return log_error_errno(errno, "Error while searching for virtiofsd: %m");
+                }
+        }
+
+        if (!virtiofsd)
+                return log_error_errno(SYNTHETIC_ERRNO(ESRCH), "Failed to find virtiofsd binary.");
+
+        r = find_executable("rm", &rm_path);
+        if (r < 0)
+                return log_error_errno(r, "Failed to find 'rm' binary: %m");
+
+        r = mkdtemp_malloc("/tmp/vmspawn-virtiofsd-XXXXXX", &state);
+        if (r < 0)
+                return r;
+
+        if (asprintf(&sock_name, "sock-%"PRIx64, random_u64()) < 0)
+                return log_oom();
+
+        sock_path = strjoin(state, "/", sock_name);
+        if (!sock_path)
+                return log_oom();
+
+        cmdline = strv_new(virtiofsd, "--shared-dir", directory, "--xattr", "--posix-acl", "--fd", "3");
+        if (!cmdline)
+                return log_oom();
+
+        cleanup = strv_new(rm_path, "-rf", state);
+        if (!cleanup)
+                return log_oom();
+
+        unit_name = strjoin(scope, "-virtiofsd", directory);
+        if (!unit_name)
+                return log_oom();
+
```

Comment:
smae here, if there's only one parameter, maybe make this a regular templated services, rather than a transient one

Reply:
I'm not sure this is possible as it needs to know both the scope and the directory meaning there is more than one thing to be templated?

Reply:
can't the dir be derived from %i?

Reply:
I don't see how.
because of `--bind` we need multiple virtiofsd instances in the same scope.
@richardmaw-codethink has also pointed out that this limits you to only mounting a directory once with virtiofsd, preventing multiple VMs from sharing a read only directory image.

Reply:
Another thing to note is that this limits the lengths of paths which can be used, `UNIT_NAME_MAX` (255) is considerably less than `PATH_MAX` (4096).